packages/daemon/DESIGN.md

```diff
@@ -0,0 +1,215 @@
+
+Endo is a user agent for running guest programs.
+These programs receive capabilities from the user in the form of references to
+potentially powerful local or remote objects.
+Endo uses a system called Eventual Send that that stretches references to
+remote objects over one or more capability transfer protocols (CapTP).
+Endo is an example of the Powerbox Pattern[^powerbox].
+
+[^powerbox]: See _The Powerbox Pattern_ from [How Emily Tamed the
+    Caml][powerbox].
+
+[powerbox]: https://www.hpl.hp.com/techreports/2006/HPL-2006-116.html
+
+A Promise represents a result from either the past or the future.
+Eventual Send displaces a result in both space and time.
+That is, a program can use eventual send to uniformly interact with a result
+regardless of whether the eventual result is local or remote, pending or
+settled.
+Furthermore, Eventual Send can pipeline[^pipeline] messages to an eventual result,
+like invoking a method on the result of a prior method invocation, to avoid
+unnecessary round trips through the underlying CapTP connection.
+
+[^pipeline]: See [Mark's explanation of pipelines][pipeline-erights], the
+    [Cap'n Proto explanation][pipeline-capnproto], [Wikipedia][pipeline-wikipedia], or
+    the [ECMAScript standard proposal][pipeline-proposal].
+
+[pipeline-erights]: http://erights.org/elib/distrib/pipeline.html
+[pipeline-capnproto]: https://capnproto.org/rpc.html
+[pipeline-wikipedia]: https://en.wikipedia.org/wiki/Futures_and_promises#Promise_pipelining
+[pipeline-proposal]: https://www.proposals.es/proposals/Support%20for%20Distributed%20Promise%20Pipelining
+
+Much hinges on what happens when a program restarts.
+Like most processes, Endo workers start without any memory from their previous
+run.
+So, Endo programs must manually persist[^persistence] anything they wish to
+remember after a restart.
+
+[^persistence]: Aside: Manual persistence is as opposed to Orthogonal
+    Persistence, where a program resumes exactly where it left off when it
+    restarts, albeit it by replaying every message it recieved since the last
+    snapshot or inception.  The architecture of Endo allows for Orthogonal
+    Persistence to be built on top of a manually persisted worker and a durable
+    reference to external storage for snapshots, message log, and virtual
+    objects.
+
+Endo applications ultimately receive all their data and powers from the Endo
+daemon with the user's consent.
+Endo persists **data** between restarts with a content address store in the
+user's home directory.
```

Comment:
We need to beware of a terminology ambiguity that's gonna bite us. In most contexts, when we say "Endo", we mean the endo repository, or the set of packages defined by that repo, or the level of abstraction brought about by those packages. When we talk about an "Endo application" I think we mean an application built on those packages.

Here, when you say "Endo", you seem to mean the same thing you're also calling "the Endo daemon". When you say "Endo applications", you state that you mean something with the special relationship to the Endo daemon that this doc is about to explain. I think this shift in meaning of "Endo application" is probably fine. In E terms, perhaps your "Endo application" is more like a capDesk caplet than like an E application.

But repurposing "Endo" to mean specifically the Endo daemon would be painful. OTOH, "the Endo daemon" is a exactly the kind of mouthful that drives people to synecdoche. Perhaps we need a new term?

Reply:
I’ll review this to make sure Endo Daemon is explicit. If I say `endo`, I mean the CLI for the Endo Daemon.

I would like to entertain “Endo caplet” or “Endo extension” since “Endo application” might invite ambiguity with a GUI Application.

I don’t know the distinction between E application and capDesk caplet.

In the reactions, cc @dckc @michaelfig 
- 🎉 means Endo extension
- 🚀 means Endo caplet
😉

Reply:
> I don’t know the distinction between E application and capDesk caplet.

We should talk about this verbally sometime. It is interesting and relevant.